### ISSUE TITLE
Customizable source directories for `gleam build`

### ISSUE THREAD

**Issue by aslilac** [gleam-lang/gleam#1556]
Based on a [conversation in Discord](https://discord.com/channels/768594524158427167/768594524158427170/951658093245263913)

./src/ is a great default, but occasionally you run into constraints that require something a bit more customizable.

In gleam.toml, we could potentially let you specify something like...
```toml
source = "./api/"
```

We might even want to let you specify multiple source directories...
```toml
source = ["./api/", "./lib/"]
```
...which would unfortunately be a bit ambiguous. Should `import a/b` resolve to ./api/a/b.gleam or ./lib/a/b.gleam if there's a conflict? Maybe we just throw an error if it could be resolved more than one way?

Another solution could be a "shared root" similar to what [TypeScript does](https://www.typescriptlang.org/tsconfig#rootDir)
```toml
# Must be a parent of all modules
sourceRoot = "./" 
# These paths should probably 
modules = ["api", "lib"] # These would probably be "Gleam paths" and not fs paths.
```

...which in Gleam could then allow you to...

```gleam
import lib/some/shared/logic.{Whatever}

...
```

...where the prefix makes imports unambiguous.

**Comment by TanklesXL** [comment]
I like the idea of configurable source directory but would be worried about the impacts of the shared-root idea. If gleam ever allows for local dependencies, does the need for shared-root style setups disappear? If the application has a `main`, its location becomes ambiguous, do we then need another config param to specify which to directory to use as the entrypoint of the application?

**Comment by aslilac** [comment]
> Would it be possible to achieve the goal with a single customisable source path?

Yes, but I just keep thinking about [what you said here](https://discord.com/channels/768594524158427167/811254431051808778/950801636547239986) :\

Multiple directories would ultimately just be a "nice to have". Super helpful, but not strictly necessary. A customizable source directory would be enough to at least finish out a prototype.

**Comment by lpil** [comment]
> I like the idea of configurable source directory but would be worried about the impacts of the shared-root idea. If gleam ever allows for local dependencies, does the need for shared-root style setups disappear? 

This is a really good observation. I think it might, yes.

> If the application has a main, its location becomes ambiguous, do we then need another config param to specify which to directory to use as the entrypoint of the application?

We would collect all the srcs into one tree and error if there are any duplicates from the different roots. Once that is done there can only every be 1 or 0 entrypoint modules.
